In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('customer_analytics_dataset.csv')
data.head(10)

,customer_id,age,gender,country,avg_order_value,total_orders,last_purchase,is_fraudulent,preferred_category,email_open_rate,customer_since,loyalty_score,churn_risk
0,CUST_8270,30,Female,Brazil,101.08,8,176,1,Beauty,25.6,2024-06-05,50,0.20
1,CUST_1860,53,Female,USA,90.39,10,88,0,Electronics,12.3,2024-02-19,37,0.34
2,CUST_6390,73,Male,Australia,83.28,6,203,0,Sports,NaN,2024-04-16,65,0.05
3,CUST_6191,30,Other,Japan,109.90,9,346,1,Electronics,42.9,2020-07-08,93,0.19
4,CUST_6734,29,Female,Canada,269.38,16,342,0,Fashion,5.3,2025-04-09,79,0.15
5,CUST_7265,54,Male,Canada,114.31,11,210,0,Beauty,56.5,2021-10-02,44,0.21
6,CUST_1466,27,Female,Canada,34.03,7,212,0,Electronics,0.2,2021-12-19,60,0.39
7,CUST_5426,77,Female,Brazil,171.25,9,20,0,Electronics,15.1,2021-02-09,29,0.24
8,CUST_6578,73,Female,Australia,116.79,10,87,0,Sports,14.7,2020-09-02,85,0.37
9,CUST_9322,26,Male,France,55.93,4,12,0,Fashion,68.8,2024-02-25,63,0.52


In [3]:
data.shape

(5000, 13)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   customer_id         5000 non-null   object 
 1   age                 5000 non-null   int64  
 2   gender              5000 non-null   object 
 3   country             5000 non-null   object 
 4   avg_order_value     4750 non-null   float64
 5   total_orders        5000 non-null   int64  
 6   last_purchase       5000 non-null   int64  
 7   is_fraudulent       5000 non-null   int64  
 8   preferred_category  5000 non-null   object 
 9   email_open_rate     4750 non-null   float64
 10  customer_since      5000 non-null   object 
 11  loyalty_score       5000 non-null   int64  
 12  churn_risk          5000 non-null   float64
dtypes: float64(3), int64(5), object(5)
memory usage: 507.9+ KB


In [5]:
data.isnull().sum()

customer_id             0
age                     0
gender                  0
country                 0
avg_order_value       250
total_orders            0
last_purchase           0
is_fraudulent           0
preferred_category      0
email_open_rate       250
customer_since          0
loyalty_score           0
churn_risk              0
dtype: int64

In [6]:
data.head(2)

,customer_id,age,gender,country,avg_order_value,total_orders,last_purchase,is_fraudulent,preferred_category,email_open_rate,customer_since,loyalty_score,churn_risk
0,CUST_8270,30,Female,Brazil,101.08,8,176,1,Beauty,25.6,2024-06-05,50,0.20
1,CUST_1860,53,Female,USA,90.39,10,88,0,Electronics,12.3,2024-02-19,37,0.34


In [7]:
data.describe()

,age,avg_order_value,total_orders,last_purchase,is_fraudulent,email_open_rate,loyalty_score,churn_risk
count,5000.000000,4750.000000,5000.000000,5000.000000,5000.000000,4750.000000,5000.000000,5000.000000
mean,48.163200,108.442857,10.027000,180.073200,0.025800,50.714842,50.039400,0.284484
std,17.880797,69.265559,3.163838,104.926518,0.158554,29.098706,28.832151,0.159690
min,18.000000,10.660000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000
25%,33.000000,57.805000,8.000000,89.000000,0.000000,25.225000,25.000000,0.160000
50%,48.000000,93.190000,10.000000,178.000000,0.000000,50.950000,50.000000,0.260000
75%,64.000000,142.197500,12.000000,270.000000,0.000000,76.800000,75.000000,0.390000
max,79.000000,555.460000,23.000000,364.000000,1.000000,100.000000,99.000000,0.900000


In [8]:
data.duplicated().value_counts()


False    5000
Name: count, dtype: int64

In [9]:

# columns drop krne ki koshish kr rahha tha but 
# mujhay lagta hai k isme ziada data loss hojae ga
# data.drop(columns=['customer_id', ])

### dateTime column main convert karna zaroori hai so that pata chal sake k customer kitne time sy hamary sath hai

In [10]:
data['customer_since'] = pd.to_datetime(data['customer_since'])

# ispr time series analysis kr k dekhte lete hain
data['customer_tenure_days'] = (pd.to_datetime('today') - data['customer_since']).dt.days

In [11]:
data = data.drop(columns=['customer_since'])

In [12]:
data.head(2)

,customer_id,age,gender,country,avg_order_value,total_orders,last_purchase,is_fraudulent,preferred_category,email_open_rate,loyalty_score,churn_risk,customer_tenure_days
0,CUST_8270,30,Female,Brazil,101.08,8,176,1,Beauty,25.6,50,0.20,800
1,CUST_1860,53,Female,USA,90.39,10,88,0,Electronics,12.3,37,0.34,907


In [13]:
data.duplicated(subset=['customer_id']).sum()

np.int64(1191)

### Un_use_full columns ko drop is lea kr raha hun k data sy noise kam ho jae

In [14]:
data.drop(columns=['customer_id']).head(5)

,age,gender,country,avg_order_value,total_orders,last_purchase,is_fraudulent,preferred_category,email_open_rate,loyalty_score,churn_risk,customer_tenure_days
0,30,Female,Brazil,101.08,8,176,1,Beauty,25.6,50,0.20,800
1,53,Female,USA,90.39,10,88,0,Electronics,12.3,37,0.34,907
2,73,Male,Australia,83.28,6,203,0,Sports,NaN,65,0.05,850
3,30,Other,Japan,109.90,9,346,1,Electronics,42.9,93,0.19,2228
4,29,Female,Canada,269.38,16,342,0,Fashion,5.3,79,0.15,492


### Kuch Missing Values hain unko theek kr raha hun
- avg_order_value (250) float type
- email_open_rate (250) float type

In [15]:
data['avg_order_value'] = data['avg_order_value'].fillna(data['avg_order_value'].median())

In [16]:
data['email_open_rate'] = data['email_open_rate'].fillna(data['email_open_rate'].median())

In [21]:
print(data.info())
print(data.isnull().sum())
print(data.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customer_id           5000 non-null   object 
 1   age                   5000 non-null   int64  
 2   gender                5000 non-null   object 
 3   country               5000 non-null   object 
 4   avg_order_value       5000 non-null   float64
 5   total_orders          5000 non-null   int64  
 6   last_purchase         5000 non-null   int64  
 7   is_fraudulent         5000 non-null   int64  
 8   preferred_category    5000 non-null   object 
 9   email_open_rate       5000 non-null   float64
 10  loyalty_score         5000 non-null   int64  
 11  churn_risk            5000 non-null   float64
 12  customer_tenure_days  5000 non-null   int64  
dtypes: float64(3), int64(6), object(4)
memory usage: 507.9+ KB
None
customer_id             0
age                     0
gender 

In [20]:
data.head(5)

,customer_id,age,gender,country,avg_order_value,total_orders,last_purchase,is_fraudulent,preferred_category,email_open_rate,loyalty_score,churn_risk,customer_tenure_days
0,CUST_8270,30,Female,Brazil,101.08,8,176,1,Beauty,25.60,50,0.20,800
1,CUST_1860,53,Female,USA,90.39,10,88,0,Electronics,12.30,37,0.34,907
2,CUST_6390,73,Male,Australia,83.28,6,203,0,Sports,50.95,65,0.05,850
3,CUST_6191,30,Other,Japan,109.90,9,346,1,Electronics,42.90,93,0.19,2228
4,CUST_6734,29,Female,Canada,269.38,16,342,0,Fashion,5.30,79,0.15,492


### Mere hisab sy ab data_almost good ho gaya hai, isme main yeh jo clean data hy isko eik new file main export kr raha hun model testing k lea 

In [23]:
# data.to_csv('fraud_data_cleanned.csv', index = False)

### ab is clearned data ko load kar k (encoding aur model train karte hain)

In [24]:
df = pd.read_csv('fraud_data_cleanned.csv')

In [26]:
df.head(4)

,customer_id,age,gender,country,avg_order_value,total_orders,last_purchase,is_fraudulent,preferred_category,email_open_rate,loyalty_score,churn_risk,customer_tenure_days
0,CUST_8270,30,Female,Brazil,101.08,8,176,1,Beauty,25.60,50,0.20,800
1,CUST_1860,53,Female,USA,90.39,10,88,0,Electronics,12.30,37,0.34,907
2,CUST_6390,73,Male,Australia,83.28,6,203,0,Sports,50.95,65,0.05,850
3,CUST_6191,30,Other,Japan,109.90,9,346,1,Electronics,42.90,93,0.19,2228


In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customer_id           5000 non-null   object 
 1   age                   5000 non-null   int64  
 2   gender                5000 non-null   object 
 3   country               5000 non-null   object 
 4   avg_order_value       5000 non-null   float64
 5   total_orders          5000 non-null   int64  
 6   last_purchase         5000 non-null   int64  
 7   is_fraudulent         5000 non-null   int64  
 8   preferred_category    5000 non-null   object 
 9   email_open_rate       5000 non-null   float64
 10  loyalty_score         5000 non-null   int64  
 11  churn_risk            5000 non-null   float64
 12  customer_tenure_days  5000 non-null   int64  
dtypes: float64(3), int64(6), object(4)
memory usage: 507.9+ KB


In [28]:
df_copy = df

In [29]:
num_col =  [df_copy.select_dtypes('int64', 'float64')]
cat_col = [df_copy.select_dtypes('object')]

In [34]:
np.hstack(num_col)

array([[  30,    8,  176,    1,   50,  800],
       [  53,   10,   88,    0,   37,  907],
       [  73,    6,  203,    0,   65,  850],
       ...,
       [  79,   16,   84,    0,   56, 1370],
       [  62,    8,  254,    0,   47, 1173],
       [  38,    6,  326,    0,   77, 1807]], shape=(5000, 6))

from sklearn.model_selection import train_test_split

from sklearn.ensemble import (
RandomForestClassifier, 
ExtraTreesClassifier,
GradientBoostingClassifier,
AdaBoostClassifier
)

from sklearn.tree import DecisionTreeClassifier

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder
from imblearn import 